### PR TITLE
Allow asm installer to prepare VM setup

### PR DESCRIPTION
**Pull request by jasonwzm**
* Add `vm` option and its validation
* Register cluster to Anthos environ
* Install eastwest gateway
* Expose Istiod
* Register Google Identity Provider

Testing verifies cluster is registered and identity provider is registered, and makes sure the demo app still works with the VM setup.

/cc @maasen

### CODE REVIEW

File: scripts/asm-installer/install_asm
Comment (zerobfd) [inline scripts/asm-installer/install_asm:228]: Please gate this on enable-gcp-components and otherwise fail if they're not installed

File: scripts/asm-installer/install_asm
Comment (zerobfd) [inline scripts/asm-installer/install_asm:563]: I think "register a cluster to an environ" is a bit clearer

File: scripts/asm-installer/install_asm
Comment (zerobfd) [inline scripts/asm-installer/install_asm:101]: Can you name it USE_ to be the same as Hub WIP? USE_VMS maybe

File: scripts/asm-installer/install_asm
Comment (zerobfd) [reply scripts/asm-installer/install_asm:228]: In fact, why don't you move this to the block above with init_meshconfig, init_gcp_cas, etc. and then it will be taken care of

File: scripts/asm-installer/install_asm
Comment (zerobfd) [inline scripts/asm-installer/install_asm:849]: Is there a reason we can't just automatically do this instead of making the user re-run the script?

File: scripts/asm-installer/install_asm
Comment (zerobfd) [inline scripts/asm-installer/install_asm:1911]: Please use "${PACKAGE_DIRECTORY}/expansion/expose-istiod.yaml"

File: scripts/asm-installer/tests/run_basic_suite_vm
Comment (zerobfd) [inline scripts/asm-installer/tests/run_basic_suite_vm:87]: Specify "registered to environ"

**Review by zerobfd — CHANGES_REQUESTED** [review]

File: scripts/asm-installer/install_asm
Comment (maasen) [inline scripts/asm-installer/install_asm:1640]: Why are we not using the gcloud cmd for this?

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm]: Done. Changed the CRD name to be more specific (also did the same to the other membership check below).

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm:1663]: This enables WI on the environ. This is not at cluster level and is gated by --enable-registration.

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm]: I made GKE_CLUSTER to use GKE_URI and put it at the beginning of the script. I don't want to put MEMBERSHIP_NAME in a separate function. The other functions that need to use MEMBERSHIP_NAME should use the result from get hub workload identity function.

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm]: This is not enabling a GCP component. It is just registering a CR.

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm]: Done.

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm]: Done. Renamed to USE_VM. VMS sometimes has other representation.

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm]: Done.

File: scripts/asm-installer/tests/run_basic_suite_vm
Comment (jasonwzm) [reply scripts/asm-installer/tests/run_basic_suite_vm]: Done.

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm]: We need to know the membership id of the cluster in order to retrieve it deterministically.

**Review by zerobfd — COMMENTED** [review]
Looks good, after these two issues are resolved I'll approve as long as you tested all of the cases locally to double check.

File: scripts/asm-installer/install_asm
Comment (jasonwzm) [reply scripts/asm-installer/install_asm:565]: Done.

**Comment by jasonwzm** [comment]
> How do you handle the case where someone has already registered the cluster to an Environ and not enabled workload identity?

@zerobfd That should be handled by hub-meshca option. Currently we don't validate that case and we basically assume users to have WI. I can create a separate PR to improve that.

**Review by zerobfd — APPROVED** [review]
We can merge when tests pass.

**Comment by jasonwzm** [comment]
The test pipeline keeps failing due to cluster creation. VM test succeeded: https://console.cloud.google.com/cloud-build/builds;region=global/8bf40d5e-64dd-4577-8e0a-61d14dd3cd6a;step=9?project=asm-scriptaro-oss

@zerobfd Can we just merge?

**Comment by zerobfd** [comment]
The latest run has a legitimate error in step 8.

**Comment by jasonwzm** [comment]
@zerobfd I added function to generate membership name to truncate the name and avoid conflict. PTAL.

**Comment by jasonwzm** [comment]
Also clean up memberships at the end of the tests. Need to manually clean the memberships before tests can pass.

**Comment by jasonwzm** [comment]
The vm test and a non-vm test passed locally.

**Pull request merged**
